src/entities/permit/index.ts

```diff
@@ -0,0 +1,130 @@
+import { Client, PublicActions, WalletActions } from 'viem';
+
+import { weightedPoolAbi_V3 } from '@/abi';
+import { Hex } from '@/types';
+import { BALANCER_ROUTER, MAX_UINT256 } from '@/utils';
+import { getNonce } from './helper';
+import { RemoveLiquidityBaseBuildCallInput } from '../removeLiquidity/types';
+import { getAmountsCall } from '../removeLiquidity/helper';
+
+export type PermitSignature = {
+    r: Hex;
+    s: Hex;
+    v: number;
+};
+
+export type PermitApproval = {
+    /** Address of the token to approve */
+    token: Hex;
+    /** Owner of the tokens. Usually the currently connected address. */
+    owner: Hex;
+    /** Address to grant allowance to */
+    spender: Hex;
+    /** Amount to approve */
+    amount: bigint;
+    /** Nonce of the permit */
+    nonce: bigint;
+    /** Expiration of this approval, in SECONDS */
+    deadline: bigint;
+};
+
+export type Permit = {
+    batch: PermitApproval[];
+    signatures: Hex[];
+};
+
+export class PermitHelper {
+    static signRemoveLiquidityApproval = async (
+        input: RemoveLiquidityBaseBuildCallInput & {
+            client: Client & WalletActions & PublicActions;
+            owner: Hex;
+        },
+    ): Promise<Permit> => {
+        const amounts = getAmountsCall(input);
+        const { permitApproval, permitSignature } = await signPermit(
+            input.client,
+            input.bptIn.token.address,
+            input.owner,
+            BALANCER_ROUTER[input.chainId],
+            amounts.maxBptAmountIn,
+        );
+        return { batch: [permitApproval], signatures: [permitSignature] };
+    };
+}
+
+/**
+ * Signs a permit for a given ERC-2612 ERC20 token using the specified parameters.
+ *
+ * @param { Client & WalletActions & PublicActions } client - Wallet client to invoke for signing the permit message
+ */
+const signPermit = async (
+    client: Client & WalletActions & PublicActions,
+    token: Hex,
+    owner: Hex,
+    spender: Hex,
+    amount = MAX_UINT256,
+    nonce?: bigint,
```

Comment:
I think we can tidy this up a bit. Doesn't look like we ever expect nonce to be passed (e.g. in signRemoveLiquidityApproval). 
I think we should:
* move `getNonce` out of `signPermit` and into `signRemoveLiquidityApproval`
then either:
* change `signPermit` nonce to be non-optional or
* change `signRemoveLiquidityApproval` to accept an optional nonce

Reply:
Yeap, that makes sense 👍

Reply:
A natural extension of this logic is to allow for `nonce` and `deadline` to be provided as input on permit2. The implementation is a bit more complex because user has to provide these configs for each tokenIn. I tried to keep it as flexible as possible. Let me know what you think.